From 7310234391ec9842bed36a91013156327591d05d Mon Sep 17 00:00:00 2001
From: Jack Ping CHNG <jack.ping.chng@linux.intel.com>
Date: Mon, 20 Apr 2020 17:34:30 +0800
Subject: [PATCH] DRV_DPM_SW-246: DPM remove dp_dev_list_free and free dp_dev

- removed dp_dev_list_free as it is not in used
- free dp_dev to avoid memory leak
- make sure that net_device_ops is not overwritten for GRX500
---
 drivers/net/datapath/dpm/datapath_instance.c | 32 +++++-----------------------
 1 file changed, 5 insertions(+), 27 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_instance.c b/drivers/net/datapath/dpm/datapath_instance.c
index b07294f7ebac..e95670fd0473 100644
--- a/drivers/net/datapath/dpm/datapath_instance.c
+++ b/drivers/net/datapath/dpm/datapath_instance.c
@@ -14,14 +14,8 @@
 int dp_cap_num;
 struct dp_hw_cap hw_cap_list[DP_MAX_HW_CAP];
 
-/*Subif DEV hash list
- *Note: in current implementation, its memory will not be really removed,
- *Instead, it is just move to a free hash list dp_dev_list_free
- *Reason: if we really free the memory, net_device's opt callback sometimes
- *will be set to NULL?? Need further check
- */
+/*Subif DEV hash list */
 struct hlist_head dp_dev_list[DP_DEV_HASH_SIZE];
-struct hlist_head dp_dev_list_free[DP_DEV_HASH_SIZE];
 
 /*Module hash list */
 struct hlist_head dp_mod_list[DP_MOD_HASH_SIZE];
@@ -264,16 +258,6 @@ int dp_inst_add_dev(struct net_device *dev, char *subif_name, int inst,
 
 	subif->subif = ctp;
 	dp_dev = dp_dev_lookup(&dp_dev_list[idx], dev, subif_name, flag);
-
-	if (!dp_dev) { /*search free list */
-		dp_dev = dp_dev_lookup(&dp_dev_list_free[idx],
-				       dev, subif_name, flag);
-		if (dp_dev) {
-			hlist_del(&dp_dev->hlist); /*remove from free list */
-			dp_dev->count = 0;
-			new_f = 1;
-		}
-	}
 	if (!dp_dev) { /*alloc new */
 		dp_dev = devm_kzalloc(&g_dp_dev->dev, sizeof(*dp_dev),
 				      GFP_ATOMIC);
@@ -300,7 +284,7 @@ int dp_inst_add_dev(struct net_device *dev, char *subif_name, int inst,
 				sizeof(dp_dev->subif_name) - 1);
 		}
 		hlist_add_head(&dp_dev->hlist, &dp_dev_list[idx]);
-#if IS_ENABLED(CONFIG_PPA) || IS_ENABLED(CONFIG_QOS_MGR)
+#if (IS_ENABLED(CONFIG_PPA) || IS_ENABLED(CONFIG_QOS_MGR)) && !IS_ENABLED(CONFIG_GRX500_CBM)
 		/*backup ops*/
 		if (dev) {
 			dev->features |= NETIF_F_HW_TC;
@@ -370,15 +354,11 @@ int dp_inst_del_dev(struct net_device *dev, char *subif_name, int inst,
 			list_del(&tmp->list);
 			dp_dev->count--;
 
-			if (!dp_dev->count) { /*move to free list */
+			if (!dp_dev->count) {
 				hlist_del(&dp_dev->hlist);
 				if (dev)
 					dp_ops_reset(dp_dev, dev);
-				/*do't really free now
-				 *in case network stack is holding the callback
-				 */
-				hlist_add_head(&dp_dev->hlist,
-					       &dp_dev_list_free[idx]);
+				devm_kfree(&g_dp_dev->dev, dp_dev);
 			}
 			return 0;
 		}
@@ -678,10 +658,8 @@ int dp_inst_init(u32 flag)
 
 	dp_cap_num = 0;
 	memset(hw_cap_list, 0, sizeof(hw_cap_list));
-	for (i = 0; i < DP_DEV_HASH_SIZE; i++) {
+	for (i = 0; i < DP_DEV_HASH_SIZE; i++)
 		INIT_HLIST_HEAD(&dp_dev_list[i]);
-		INIT_HLIST_HEAD(&dp_dev_list_free[i]);
-	}
 
 	return 0;
 }
